beamer/agent/middleware.py

```diff
@@ -11,10 +11,36 @@
 import structlog
 from web3 import HTTPProvider, Web3
 from web3.types import RPCEndpoint, RPCResponse
+from beamer.agent.typing import ChainId
+
 
 log = structlog.get_logger(__name__)
 
 
+class _BlockCache:
+    def __init__(self) -> None:
+        self._block_cache = lru.LRU(1000)
+        self._latest_block_number = -1
+        self._latest_key: tuple[str, bool] | None = None
+        self._lock = threading.Lock()
+
+    def add_block(self, key: tuple[str, bool], data: RPCResponse) -> None:
+        with self._lock:
+            self._block_cache[key] = data
+            if self._latest_block_number < data["result"].number:
+                self._latest_key = key
+                self._latest_block_number = data["result"].number
+
+    def get_block(self, key: tuple[str, bool]) -> RPCResponse:
+        return self._block_cache.get(key)
+
+    def get_latest_block(self) -> RPCResponse:
```

Comment:
the return value should be RPCResponse | None

because self._latest_key can be initialized with None